src/Experimental/ShortestPaths/ShortestPaths.jl

```diff
@@ -0,0 +1,212 @@
+
+using LightGraphs: dijkstra_shortest_paths, bellman_ford_shortest_paths, floyd_warshall_shortest_paths,
+    desopo_pape_shortest_paths, a_star
+import Base:convert, getproperty
+import LightGraphs: enumerate_paths
+
+const paths = enumerate_paths
+
+struct LGEnvironment
+    threaded::Bool
+    parallel::Bool
+    LGEnvironment() = new(false, false)
+end
+     
+abstract type AbstractGraphResults end
+abstract type AbstractGraphAlgorithm end
+
+abstract type ShortestPathResults <: AbstractGraphResults end
+abstract type ShortestPathAlgorithm <: AbstractGraphAlgorithm end
+
+##################
+##   Dijkstra   ##
+##################
+struct DijkstraResults{T<:Real, U<:Integer}  <: ShortestPathResults
+    parents::Vector{U}
+    dists::Vector{T}
+    predecessors::Vector{Vector{U}}
+    pathcounts::Vector{UInt64}
+    closest_vertices::Vector{U}
+end
+
+convert(::Type{AbstractPathState}, spr::DijkstraResults) = convert(DijkstraState, spr)
+convert(::Type{<:DijkstraResults}, s::DijkstraState) =
+    DijkstraResults(s.parents, s.dists, s.predecessors, s.pathcounts, s.closest_vertices)
+
+convert(::Type{<:DijkstraState}, spr::DijkstraResults) =
+    DijkstraState(spr.parents, spr.dists, spr.predecessors, spr.pathcounts, spr.closest_vertices)
+
+# These constructors do not copy. They probably should, but you shouldn't be changing the values in any case.
+# Can we just document that it's undefined behavior to change any element of a SPR?
+DijkstraResults(s::DijkstraState) = convert(DijkstraResults, s)
+DijkstraState(spr::DijkstraResults) = convert(DijkstraState, spr)
+
+struct Dijkstra <: ShortestPathAlgorithm
+    all_paths::Bool
+    track_vertices::Bool
+
+    Dijkstra() = new(false, false)
+end
+
+
+##################
+##    A-Star    ##
+##################
+struct AStarResults{E<:AbstractEdge} <: ShortestPathResults
+    path::Vector{E}
+end
+# for completeness and consistency.
+struct AStarState{E<:AbstractEdge} <: AbstractPathState
+    path::Vector{E}
+end
+
+function paths(s::AStarState)
+    T = eltype(eltype(s.path))
+    n = length(s.path)
+    n == 0 && return Vector{T}()
+    p = Vector{T}(undef, n+1)
+    p[1] = src(s.path[1])
+    p[end] = dst(s.path[end])
+    for i in 2:n
+        p[i] = src(s.path[i])
+    end
+    return p
+end
+
+convert(::Type{AbstractPathState}, spr::AStarResults) = convert(AStarState, spr)
+convert(::Type{<:AStarResults}, s::AStarState) = AStarResults(s.path)
+
+convert(::Type{<:AStarState}, spr::AStarResults) = AStarState(spr.path)
+
+struct AStar{F<:Function} <: ShortestPathAlgorithm
+    heuristic::F
+end
+AStarResults(s::AStarState) = convert(AStarResults, s)
+AStarState(spr::AStarResults) = convert(AStarState, spr)
+
+AStar(T::Type{<:Real}=Float64) = AStar(n -> zero(T))
+
+
+##################
+## Bellman-Ford ##
+##################
+struct BellmanFord <: ShortestPathAlgorithm end
+struct BellmanFordResults{T<:Real, U<:Integer} <: ShortestPathResults
```

Comment:
Should `Result` be pluralized here? It's rare for types. Same for `ShortestPathResults` and `AbstractGraphResults`

Reply:
It's a good question. I'm ok with the singular, but at least in my experience, "result" is used specifically when there is a single datum. (In this case, we have multiple results - either parents/dists, or paths.)

I'm happy to change it. Let's get a consensus.

Reply:
ping @jpfairbanks for consensus

Reply:
I think we want to go with the singular. I'll make the change.

Reply:
Results also does sound really strange to me. It looks as if it is a collection and we don't use plural for other collection types such as vectors. So yes, `Result` or `ResultSet` would be preferable I guess.